### PR TITLE
Laboratory work #3, Andrey Klimov - 23FPL1

### PR DESCRIPTION
**Pull request by QuietHellsPage**

### CODE REVIEW

File: lab_2_retrieval_w_bm25/main.py [hunk 1]
```diff
@@ -20,12 +19,11 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
-    if not isinstance(text, str):
+    if not isinstance(text, str) or not text:
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:22]: revert all changes in lab 2

File: lab_3_ann_retriever/main.py [hunk 2]
```diff
@@ -24,6 +23,7 @@ def save(self) -> dict:
             dict: State of the Node instance
         """
 
+
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:26]: revert

File: lab_3_ann_retriever/main.py [hunk 3]
```diff
@@ -7,8 +7,7 @@
 # pylint: disable=too-few-public-methods, too-many-arguments, duplicate-code, unused-argument
 from typing import Protocol
 
-Vector = tuple[float, ...]
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:10]: revert

File: lab_3_ann_retriever/main.py [hunk 4]
```diff
@@ -104,6 +107,15 @@ def tokenize(self, text: str) -> list[str] | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if not isinstance(text, str) or not text:
+            return None
+        text = text.lower()
+        tokenized_text = ''
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:113]: do you really need to create separate variable `tokenized_text`?

File: lab_3_ann_retriever/main.py [hunk 5]
```diff
@@ -117,6 +129,15 @@ def tokenize_documents(self, documents: list[str]) -> list[list[str]] | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if not isinstance(documents, list) or not all(isinstance(item, str) for item in documents) or not documents:
+            return None
+        doc = []
+        for text in documents:
+            processed_text = self.tokenize(text)
+            if processed_text == None:
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:137]: use `is None`

**Review by Vasilisa-Blyudova — CHANGES_REQUESTED** [review]
Good work! Don't forget to change the score in `settings.json`

**Comment by QuietHellsPage** [comment]
I am sorry, all the funcs in Vectorizer and engine were deleted because of update of my branch, so now there is only tokenizer, I'll try to recreate other functions asap, but maybe there are some ways to find some previous changes?

File: lab_3_ann_retriever/main.py [hunk 6]
```diff
@@ -49,6 +50,18 @@ def calculate_distance(query_vector: Vector, document_vector: Vector) -> float |
 
     In case of corrupt input arguments, None is returned.
     """
+    if query_vector is None or document_vector is None:
+        return None
+
+    if not query_vector or not document_vector:
+        return 0.0
+
+    # dist = 0.0
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:59]: rm

File: lab_3_ann_retriever/main.py [hunk 7]
```diff
@@ -157,6 +201,33 @@ def build(self) -> bool:
...
+        idf = calculate_idf(self._vocabulary, self._corpus)
+        if not isinstance(idf, dict):
+            return False
+        self._idf_values = idf
+
+        self._token2ind = {value: ind for ind, value in enumerate(self._vocabulary)}
+
+        return True
+
+        # self._idf_values = {}
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:227]: rm

File: lab_3_ann_retriever/main.py [hunk 8]
```diff
@@ -157,6 +201,33 @@ def build(self) -> bool:
...
+
+        voc = set()
+        for doc in self._corpus:
+            if all(isinstance(elem, str) for elem in doc):
+                voc.update(doc)
+            else:
+                return False
+        self._vocabulary = list(voc)
+        self._vocabulary.sort()
+        # if self._vocabulary is None:
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:215]: .

File: lab_3_ann_retriever/main.py [hunk 9]
```diff
@@ -220,6 +309,39 @@ def _calculate_tf_idf(self, document: list[str]) -> Vector | None:
...
+
+        return tuple((tf[elem] * self._idf_values[elem] if elem in document else 0.0 for elem in self._vocabulary))
+
+
+
+
+
+
+
+        # self.build()
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:331]: rm

File: lab_3_ann_retriever/main.py [hunk 10]
```diff
@@ -220,6 +309,39 @@ def _calculate_tf_idf(self, document: list[str]) -> Vector | None:
...
+
+        tf = {}
+
+        voc = set(self._vocabulary)
+        doc = set(document)
+
+        for elem in voc.union(doc):
+            tf[elem] = document.count(elem) / len(document)
+
+        return tuple((tf[elem] * self._idf_values[elem] if elem in document else 0.0 for elem in self._vocabulary))
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:323]: try to simplify

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Great! Go to 8!

File: lab_3_ann_retriever/main.py (hunk 2 above)
Comment (QuietHellsPage) [reply lab_3_ann_retriever/main.py:26]: Done

File: lab_3_ann_retriever/main.py (hunk 3 above)
Comment (QuietHellsPage) [reply lab_3_ann_retriever/main.py:10]: Done

File: lab_3_ann_retriever/main.py (hunk 6 above)
Comment (QuietHellsPage) [reply lab_3_ann_retriever/main.py]: Done

File: lab_3_ann_retriever/main.py (hunk 7 above)
Comment (QuietHellsPage) [reply lab_3_ann_retriever/main.py]: Done

File: lab_3_ann_retriever/main.py (hunk 8 above)
Comment (QuietHellsPage) [reply lab_3_ann_retriever/main.py]: Done

File: lab_3_ann_retriever/main.py (hunk 9 above)
Comment (QuietHellsPage) [reply lab_3_ann_retriever/main.py]: Done

File: lab_3_ann_retriever/main.py (hunk 4 above)
Comment (QuietHellsPage) [reply lab_3_ann_retriever/main.py]: I have 2 options. Second one is shorter, but u said that it is too difficult last time

File: lab_3_ann_retriever/main.py (hunk 10 above)
Comment (QuietHellsPage) [reply lab_3_ann_retriever/main.py]: Done. I've added the second option, but I don't like it and don't know why, though it works too

File: lab_2_retrieval_w_bm25/main.py (hunk 1 above)
Comment (artyomtugaryov) [reply lab_2_retrieval_w_bm25/main.py:22]: yes

File: lab_3_ann_retriever/main.py (hunk 3 above)
Comment (artyomtugaryov) [reply lab_3_ann_retriever/main.py:10]: no

File: lab_3_ann_retriever/main.py (hunk 2 above)
Comment (artyomtugaryov) [reply lab_3_ann_retriever/main.py:31]: again - I see this changes, so you are not done with it

File: lab_3_ann_retriever/main.py [hunk 11]
```diff
@@ -104,6 +116,17 @@ def tokenize(self, text: str) -> list[str] | None:
...
         """
+        if not isinstance(text, str) or not text:
+            return None
+
+        for element in text:
+            if not element.isalpha() and not element.isspace():
+                text = text.replace(element, ' ')
+        return self._remove_stop_words(text.lower().split())
+
+        # toks = [word for word in ''.join(char for char in text.lower() if char.isalpha() or char.isspace()).split()]
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:127]: remove

File: lab_3_ann_retriever/main.py [hunk 12]
```diff
@@ -195,6 +271,7 @@ def save(self, file_path: str) -> bool:
             bool: True if saved successfully, False in other case
         """
 
+
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:274]: revert

File: lab_3_ann_retriever/main.py [hunk 13]
```diff
@@ -208,6 +285,7 @@ def load(self, file_path: str) -> bool:
         In case of corrupt input arguments, False is returned.
         """
 
+
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:288]: revert

File: lab_3_ann_retriever/main.py [hunk 14]
```diff
@@ -220,6 +298,29 @@ def _calculate_tf_idf(self, document: list[str]) -> Vector | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if not isinstance(document, list) or not all(isinstance(item, str) for item in document) or not document:
+            return None
+
+        # tf = {}
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:304]: remove

File: lab_3_ann_retriever/main.py [hunk 15]
```diff
@@ -304,6 +440,17 @@ def retrieve_vectorized(self, query_vector: Vector) -> str | None:
...
         """
+        if not isinstance(query_vector, tuple) or not query_vector:
+            return None
+        if len(query_vector) != len(self._document_vectors[0]):
+            return None
+        retrieve_doc = self._calculate_knn(query_vector, self._document_vectors, 1)
+        if not isinstance(retrieve_doc, list) or not retrieve_doc:
+            return None
+        index, value = retrieve_doc[0]
+        doc = self._documents[index]
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:451]: do not create the variable

File: lab_3_ann_retriever/main.py [hunk 16]
```diff
@@ -343,6 +515,7 @@ def _dump_documents(self) -> dict:
             dict: document and document_vectors states
         """
 
+
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:518]: revert

File: lab_3_ann_retriever/main.py [hunk 17]
```diff
@@ -426,6 +606,74 @@ def build(self, vectors: list[Vector]) -> bool:
 
         In case of corrupt input arguments, False is returned.
         """
+        # if not isinstance(vectors, list) or not vectors or len(vectors) == 0:
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:609]: remove

File: lab_3_ann_retriever/main.py [hunk 18]
```diff
@@ -462,6 +715,7 @@ def load(self, state: dict) -> bool:
             bool: True is loaded successfully, False in other cases
         """
 
+
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:718]: revert

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Pull request closed without merging**